python/mrc/_pymrc/src/utils.cpp

```diff
@@ -170,8 +172,11 @@ json cast_from_pyobject(const py::object& source)
         return json(py::cast<std::string>(source));
     }
 
-    // else unsupported return null
-    return json();
+    // else unsupported return throw a type error
+    {
+        AcquireGIL gil;
+        throw py::type_error("Object is not JSON serializable");
+    }
```

Comment:
This error will be very frustrating for users in its current state. We need to report more information about the object that caused the exception. Especially since potentially deep in the object.

To this exception, we should add the typename of the object that caused the issue, a string representation of the object `py::str(source)` and the path of the object (you can get this with careful exception handlers and unwinding the recursive function).